compiler/rustc_abi/src/layout/ty.rs

```diff
@@ -269,6 +271,24 @@ impl<'a, Ty> TyAndLayout<'a, Ty> {
         Ty::is_transparent(self)
     }
 
+    /// If this method returns `true`, then this type should always have a `PassMode` of
+    /// `Indirect { on_stack: false, .. }` when being used as the argument type of a function with a
+    /// non-Rustic ABI (this is true for structs annotated with the
+    /// `#[rustc_pass_indirectly_in_non_rustic_abis]` attribute).
+    ///
+    /// This function handles transparent types automatically.
+    pub fn pass_indirectly_in_non_rustic_abis<C>(mut self, cx: &C) -> bool
+    where
+        Ty: TyAbiInterface<'a, C> + Copy,
+    {
+        while self.is_transparent()
+            && let Some((_, field)) = self.non_1zst_field(cx)
+        {
+            self = field;
+        }
+        Ty::is_pass_indirectly_in_non_rustic_abis_flag_set(self)
```

Comment:
so right now the type is only passed indirectly if the base type is passed indirectly. This just confused me for a while, I think this should also work:

```rust
#[repr(C)]
struct Vanilla(i32);

#[repr(transparent)]
#[rustc_pass_indirectly_in_non_rustic_abis]
struct Indirectly(Vanilla);
```

I.e. if any of the transparent wrappers is `rustc_pass_indirectly_in_non_rustic_abis` then the whole type should be passed indirectly.

Reply:
I think an error makes most sense -- the behavior you now implemented seems to violate the promise that repr(transparent) wrappers have the same ABI as their underlying type on *all* ABIs.

Reply:
Hmm yeah, looking at this again, given e.g. https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=7e1399219e6854e5ca5e5765cb8e6993 an error is probably better. `#[rustc_pass_indirectly_in_non_rustic_abis]` does feel morally like a repr hint.